Rebuildable system & appliance

nikstur opened the pull request:
## Description of changes

Add `system.rebuildable` flag

This flag allows the user to optionally exclude
switch-to-confguration.pl from toplevel.

This is interesting for appliance images where you don't want to re-build
the system. This flag is called `rebuildable` because the standard
interface to do this is `nixos-rebuild` which will not work anymore with
this change.

When https://github.com/NixOS/nixpkgs/pull/263295 lands, `environment.defaultPackages = mkDefault [ ];` can be removed from the minimal profile.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nikstur committed:
nixos: add system.switch.enable flag

This flag allows the user to optionally exclude
switch-to-confguration.pl from toplevel.

This is interesting for appliance images where you don't want to re-build
the system. This flag is called `rebuildable` because the standard
interface to do this is `nixos-rebuild` which will not work anymore with
this change.

blitz committed:
nixos/profiles: add image-based-appliance profile

nikstur committed:
nixos/profiles/minimal: remove some perl

roberth commented on nixos/modules/profiles/appliance.nix:
```diff
@@ -0,0 +1,25 @@
+
+  # The system cannot be rebuilt.
+  nix.enable = false;
+  system.rebuildable = false;
```
This should be factored out into `virtual-appliance.nix` which can then import this file.
Hardware appliances need an upgrade mechanism, and that would probably involve `switch`-ing.

roberth commented on nixos/modules/profiles/appliance.nix:
```diff
@@ -0,0 +1,25 @@
+  ];
+
+  # The system cannot be rebuilt.
+  nix.enable = false;
```
In hardware appliances we probably want a stripped down Nix so that it can receive a new system closure.

roberth commented on nixos/modules/system/activation/activatable-system.nix:
```diff
@@ -22,28 +22,30 @@ let
-      --subst-var-by systemd "${config.systemd.package}" \
-      --subst-var-by utillinux "${pkgs.util-linux}" \
-      ;
+    ${lib.optionalString config.system.rebuildable ''
```
Could you move this to a new file, `switchable-system.nix`?

nikstur replied:
We would need a new variable (e.g. `system.build.systemBuilderCommands`) to communicate between the new file and the currently local `systemBuilderCommands` in activatable-system though. Is this desired?

roberth replied:
We already have `system.systemBuilderCommands`, with merging, so I think you could add to that without introducing any new options.

nikstur replied:
Sadly that breaks the separateActivationScript. So we have to introduce a new variable or keep the changes in activatable-system

roberth replied:
`separateActivationScript` is a fairly new feature, and its name does not suggest switching.
I think it'd be ok not to support that, and perhaps provide a `separateSwitchScript` in addition.

- `separateActivationScript` is suitable for building a pre-activated immutable image without leaving the activation script around.
- `separateSwitchScript` is suitable for switching and then garbage collecting the switching tools, like some tools, or perhaps even perl, which interprets the switch script.

In such a case, `separateSwitchScript` could call `separateActivationScript`, or maybe override it to add its own binary. It wouldn't need to send its guts to the activation script module anymore.

You could implement `separateSwitchScript`, or make `system.activatableSystemBuilderCommands` an `internal` option and document that it may go away, referring to this discussion.

nikstur replied:
I went with adding a comment to `system.activatableSystemBuilderCommands`. The option should already be internal because I just override the description of `options.system.systemBuilderCommands`. 

A `separateSwitchScript` might be hard to wire up because of the way `types.unspecified` merges definitions. We need to set `system.systemBuilderArgs.toplevelVar = "out";` in switchable-system when `system.activatable = false` for for switch-to-configuration to even *build* correctly. However we can only set this once. If I also set `toplevelVar` in switchable-system we get `toplevelVar = "outout" because it merges the strings. So in effect we can only set toplevelVar once.

roberth replied:
You could set it with a different priority, such as `mkDefault` to hack around that. The real solution is to fix the type to not merge strings.
Can be done later. It's internal.

roberth commented on nixos/modules/system/activation/activatable-system.nix:
```diff
@@ -60,6 +62,17 @@ in
         do, but for image based systems, this may not be needed or not be desirable.
       '';
     };
+    system.rebuildable = lib.mkOption {
```
```suggestion
    system.switch.enable = lib.mkOption {
```

roberth commented on nixos/modules/system/activation/activatable-system.nix:
```diff
@@ -60,6 +62,17 @@ in
+      type = lib.types.bool;
+      default = true;
+      description = lib.mdDoc ''
+        Exclude switch-to-confguration.pl from toplevel, rendering the system
+        unable to be re-built via `nixos-rebuild.
```
```suggestion
        Whether to include the capability to switch configurations without reboot.
        
        Disabling this makes the system unable to be reconfigured via `nixos-rebuild switch`.
```

- Describe what it does instead of why you added it
- I believe it could still be booted into a new configuration as in `nixos-rebuild boot` (e.g. a specialisation or hardware appliance updated with Nix)
- Missing backtick

nikstur replied:
For `nixos-rebuild boot` we would still need switch-to-configuration because that's what installs the bootloader (and thus the new entries pointing to the new init), right?

roberth replied:
For the bootloader it delegates almost immediately to the `installBootLoader` script and then immediately exits, but yeah you're right.
Would be nice to factor that out. That would change the interface with `nixos-rebuild`, so should be done with care. Maybe later.

roberth commented on nixos/modules/system/activation/activatable-system.nix:
```diff
@@ -60,6 +62,17 @@ in
+        Exclude switch-to-confguration.pl from toplevel, rendering the system
+        unable to be re-built via `nixos-rebuild.
+
+        This is interesting for image based appliances where you don't want to
+        dynamically re-build the system.
```
```suggestion
        This is good for image based appliances where updates are handled outside the image. Reducing features makes the image lighter and slightly more secure.
```

roberth reviewed (commented):
Really cool, but I think this needs to distinguish between general appliances and virtual appliances. I'll have some more thoughts regarding the distinction, but first I'll lazily quote Wikipedia [Software appliance](https://en.wikipedia.org/wiki/Software_appliance):

> A software appliance is a [software application](https://en.wikipedia.org/wiki/Application_software) combined with [just enough operating system (JeOS)](https://en.wikipedia.org/wiki/Just_enough_operating_system) to run optimally on industry-standard hardware (typically a [server](https://en.wikipedia.org/wiki/Server_(computing)#Hardware)) or in a [virtual machine](https://en.wikipedia.org/wiki/Virtual_machine).

This PR elaborates what is _just enough_, and it seems to get it right for virtual appliances. However, for a hardware appliance you probably want to have more drivers and some upgrade mechanism.

By default the upgrade mechanism should probably be SSH + Nix _store_. Emphasis on store because it only has to work for `root`, and does not have to come with an interpreter.

nikstur commented:
> However, for a hardware appliance you probably want to have more drivers and some upgrade mechanism.

I should've made this clearer: the update mechanism I envision is image-based A/B booting via the systemd toolchain. At the center of this is systemd-sysupdate for which I have added a module in a previous PR https://github.com/NixOS/nixpkgs/pull/243995 In this universe you would receive a completely new store image instead of only a new store closure, so Nix isn't necessary at all. 

You would build images with the repart builder which I have also already added in a previous PR: https://github.com/NixOS/nixpkgs/pull/245250

Hardware and virtual appliance in this world are the same, you can update them the same way. However with a virtual appliance, you can also just completely exchange the disk image (but then you don't get the automatic fallback from systemd-boot and systemd-bless-boot)

roberth commented:
> via the systemd toolchain

How about `systemd-sysupdate-appliance.nix` then?

I'd be happy to start with very specific profiles and consider factoring this out into common profiles only when the other use cases have profiles and the commonality is "proven".

nikstur commented:
> How about `systemd-sysupdate-appliance.nix` then?

I called it `image-based-appliance` because that's the differentiating factor between our approaches in my mind. If you do image based updates, you do not need Nix.

roberth commented:
> I called it `image-based-appliance`

If `systemd-sysupdate` doesn't require overly specific changes to the config, then that sounds great to me, and if that changes, you can always add an extra profile specifically for it.

roberth commented on nixos/tests/non-switchable-system.nix:
```diff
@@ -0,0 +1,15 @@
+  };
+
+  testScript = ''
+    machine.fail("test -e /run/current-system/bin/switch-to-configuration")
```
This assertion makes me nervous.

```suggestion
    machine.succeed("test ! -e /run/current-system/bin/switch-to-configuration")
```

roberth commented on nixos/tests/non-switchable-system.nix:
```diff
@@ -0,0 +1,15 @@
+  testScript = ''
+    machine.fail("test -e /run/current-system/bin/switch-to-configuration")
+  '';
+}
```
Nothing else to be tested?
Maybe test that your profile doesn't have `perl`? Or isn't it quite there yet?

[system.forbiddenDependenciesRegex](https://search.nixos.org/options?channel=23.05&show=system.forbiddenDependenciesRegex&from=0&size=50&sort=relevance&type=packages&query=forbidden)

nikstur replied:
Not quite there yet, but I have some more work in the pipeline ;)

roberth reviewed (commented):
Implementation lgtm.

Needs documentation and maybe promote it with a release note?

nikstur committed:
nixos: release notes for `system.switch.enable`

nikstur committed:
nixos/docs: add non-switchable-systems section

nikstur commented:
Added docs and release notes.

@ofborg test non-switchable-system switchTest

roberth commented on nixos/doc/manual/development/non-switchable-systems.section.md:
```diff
@@ -0,0 +1,9 @@
+
+If you want to build such a system, you can exclude all the tools for
+re-building from your system by setting `config.system.switch.enable = false;`.
+Removing these tools makes the image lighter and slightly more secure.
```
What about the profile?

lheckemann commented on nixos/doc/manual/development/non-switchable-systems.section.md:
```diff
@@ -0,0 +1,21 @@
+# Non Switchable Systems {#sec-non-switchable-system}
+
+In certain systems, most notably image based appliances, updates are handled
+outside the system. This means that you do not need to re-build your
```
```suggestion
outside the system. This means that you do not need to rebuild your
```
for consistency with ~everything else

lheckemann commented on nixos/doc/manual/development/non-switchable-systems.section.md:
````diff
@@ -0,0 +1,21 @@
+}
+```
+
+The most notable deviation of this profile from a standard NixOS system is that
+it disables your system from re-building itself. It sets
+`config.system.switch.enable = false;`, which excludes
+`switch-to-configuration`, the central script called by `nixos-rebuild`, from
+your system. Removing this script makes the image lighter and slightly more
+secure.
````
This feels like an odd claim -- it doesn't prevent switching to another system, only from switching _to_ the system with this option enabled. Copying a "rebuildable" system into the Nix store on the machine in question, then running its switch-to-configuration, would still work as far as I understand...

nikstur replied:
You're right of course. Changed the wording slightly. Should be more accurate now.

lheckemann commented:
`nixos-rebuild` should probably detect when it's built a system that can't be switched to, and provide a helpful error message in such a case. I'm not sure how important that is for this PR though.

nikstur commented:
> `nixos-rebuild` should probably detect when it's built a system that can't be switched to, and provide a helpful error message in such a case. I'm not sure how important that is for this PR though.

That would be nice, however, I think it's not needed for this PR.

roberth commented:
> `nixos-rebuild` should probably detect

I feel like `nixos-rebuild`, in the state it is currently in, is saturated with all sorts of complexity.
`switch-to-configuration: No such file or directory` doesn't seem all that bad for this rather specialized use case.

I would love to have such a check after `nixos-rebuild` is refactored, but until then, let's keep it simple.

emilylange commented on nixos/modules/profiles/minimal.nix:
```diff
@@ -18,6 +18,15 @@ with lib;
 
   documentation.nixos.enable = mkDefault false;
 
+  # Perl is a default package.
+  environment.defaultPackages = mkDefault [ ];
+
+  # The lessopen package pulls in Perl.
+  programs.less.lessopen = mkDefault null;
+
+  # This pulls in nixos-containers which depends on Perl.
+  boot.enableContainers = mkDefault false;
+
```
Can we please add a release note entry for `boot.enableContainers`?

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574 would have been fitting as well.

emilylange replied:
Created #269970 for a simple eval warning when containers are used but disabled (e.g. by importing the minimal profile).
